### PR TITLE
DRAFT: move some static functions to CandidateStateManager

### PR DESCRIPTION
**Pull request by ktlshya**
Issues: #2142
Hi, I don't know if this is the right way to do this , could you give me a feedback of this draft.

### CODE REVIEW

**Comment by yutongp** [comment]
from my understanding, what @CoderZhi means is in function like `readStateBuckets` instead of passing in a `StateReader`, pass in a `CandidateStateManager` or better a `CandidateStateReader`, so you don't need to do the type assertion in all following functions everywhere, but just top level functions like `ReadState`

File: action/protocol/staking/read_state.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ import (
 
 func readStateBuckets(ctx context.Context, sr protocol.StateReader,
```
Comment (CoderZhi) [inline action/protocol/staking/read_state.go:20]: as @yutongp commented, we should change "sr protocol.StateReader" to "csr CandidateStateReader" or "csm CandidateStateManager". This is the key part of the issue. With this kind of changes, you can directly call csr.allBuckets() and etc.

Comment (ktlshya) [reply action/protocol/staking/read_state.go:20]: Is that to say I don't need to change the param ```StateReader``` of ```ReadState()``` to csm or csr @CoderZhi @yutongp

File: action/protocol/staking/protocol.go [hunk 2]
```diff
@@ -331,21 +354,25 @@ func (p *Protocol) ReadState(ctx context.Context, sr protocol.StateReader, metho
 		return nil, errors.Wrap(err, "failed to unmarshal request")
 	}
 
-	center, err := getOrCreateCandCenter(sr)
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go:334]: this is not necessary for some cases, such as readStateBuckets and readStateBucketsByVoter. From which we can tell that it is not necessary to create candidate center here (which may be time consuming).

File: action/protocol/staking/read_state.go [hunk 3]
```diff
@@ -54,21 +53,21 @@ func readStateBucketsByVoter(ctx context.Context, sr protocol.StateReader,
...
-func readStateBucketsByCandidate(ctx context.Context, sr protocol.StateReader, cc CandidateCenter,
+func readStateBucketsByCandidate(ctx context.Context, csr CandidateStateReader, cc CandidateCenter,
 	req *iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate) (*iotextypes.VoteBucketList, error) {
 	c := cc.GetByName(req.GetCandName())
 	if c == nil {
 		return &iotextypes.VoteBucketList{}, nil
 	}
 
-	indices, err := getCandBucketIndices(sr, c.Owner)
+	indices, err := csr.getCandBucketIndices(c.Owner)
```
Comment (CoderZhi) [inline action/protocol/staking/read_state.go:63]: the reason we need to create candidate center outside is that we need to interact with candidate center to get candidate by name. What if candidate center is part of csr?

**Comment by ktlshya** [comment]
fork a new one

**Pull request closed without merging**